Panic with signed peer records in go-lib2p v0.9.0

We upgraded to the latest go-libp2p in the pubsub testplans in `libp2p/test-plans` (see https://github.com/libp2p/test-plans/pull/7) and now identify no longer works.

The smoking gun is this panic:
```
2020-05-19T16:32:23.759Z	ERROR	net/identify	identify/id.go:432	latest peer record does not exist. identify message incomplete!
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x6dfb97]

goroutine 234 [running]:
github.com/libp2p/go-libp2p-core/record.(*Envelope).Marshal(0x0, 0xc0005b68e0, 0x8, 0x8, 0x203000, 0x2)
	/go/pkg/mod/github.com/libp2p/go-libp2p-core@v0.5.6/record/envelope.go:196 +0x37
github.com/libp2p/go-libp2p/p2p/protocol/identify.(*IDService).populateMessage(0xc0001f1040, 0xc0007780b0, 0xf83440, 0xc00048a720, 0xc000354380)
	/go/pkg/mod/github.com/libp2p/go-libp2p@v0.9.0/p2p/protocol/identify/id.go:469 +0x4ca
github.com/libp2p/go-libp2p/p2p/protocol/identify.(*IDService).sendIdentifyResp(0xc0001f1040, 0xf83f40, 0xc0003129c0)
	/go/pkg/mod/github.com/libp2p/go-libp2p@v0.9.0/p2p/protocol/identify/id.go:402 +0x39a
github.com/libp2p/go-libp2p/p2p/host/basic.(*BasicHost).SetStreamHandler.func1(0xc0001fe280, 0xe, 0x7f0c590e5198, 0xc0003129c0, 0x1, 0x0)
	/go/pkg/mod/github.com/libp2p/go-libp2p@v0.9.0/p2p/host/basic/basic_host.go:476 +0x9d
created by github.com/libp2p/go-libp2p/p2p/host/basic.(*BasicHost).newStreamHandler
	/go/pkg/mod/github.com/libp2p/go-libp2p@v0.9.0/p2p/host/basic/basic_host.go:327 +0x63c
```

--- a/p2p/protocol/identify/id.go
+++ b/p2p/protocol/identify/id.go
@@ -428,9 +428,6 @@ func (ids *IDService) getSnapshot() *identifySnapshot {
 	if !ids.disableSignedPeerRecord {
 		if cab, ok := peerstore.GetCertifiedAddrBook(ids.Host.Peerstore()); ok {
 			snapshot.record = cab.GetPeerRecord(ids.Host.ID())
-			if snapshot.record == nil {
-				log.Errorf("latest peer record does not exist. identify message incomplete!")
-			}
 		}
 	}
 	snapshot.addrs = ids.Host.Addrs()
@@ -465,7 +462,7 @@ func (ids *IDService) populateMessage(
 		mes.ListenAddrs = append(mes.ListenAddrs, addr.Bytes())
 	}
 
-	if !ids.disableSignedPeerRecord {
+	if !ids.disableSignedPeerRecord && snapshot.record != nil {
 		recBytes, err := snapshot.record.Marshal()
 		if err != nil {
 			log.Errorf("error marshaling peer record: %v", err)

--- a/p2p/protocol/identify/id_test.go
+++ b/p2p/protocol/identify/id_test.go
@@ -686,6 +686,37 @@ func TestUserAgent(t *testing.T) {
 	}
 }
 
+func TestNotListening(t *testing.T) {
+	// Make sure we don't panic if we're not listening on any addresses.
+	//
+	// https://github.com/libp2p/go-libp2p/issues/939
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	h1, err := libp2p.New(
+		ctx,
+		libp2p.NoListenAddrs,
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer h1.Close()
+
+	h2, err := libp2p.New(
+		ctx,
+		libp2p.ListenAddrStrings("/ip4/127.0.0.1/tcp/0"),
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer h2.Close()
+
+	err = h1.Connect(ctx, peer.AddrInfo{ID: h2.ID(), Addrs: h2.Addrs()})
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
 func TestSendPushIfDeltaNotSupported(t *testing.T) {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()

